synapse/lib/lmdbslab.py

```diff
@@ -363,8 +393,109 @@ def _growMapSize(self, size=None):
         self.lenv.set_mapsize(mapsize)
         self.mapsize = mapsize
 
+        self.resizeevent.set()
+
         return self.mapsize
 
+    def _memorylockloop(self):
+        '''
+        Separate thread loop that manages the prefaulting and locking of the memory backing the data file
+        '''
+        if not s_thishost.get('hasmemlocking'):
+            return
+        MAX_TOTAL_PERCENT = .90  # how much of all the RAM to take
+        MAX_LOCK_AT_ONCE = s_const.gibibyte
+
+        # Calculate a reasonable maximum amount of memory to lock
+
+        s_thisplat.maximizeMaxLockedMemory()
+        locked_ulimit = s_thisplat.getMaxLockedMemory()
+        if locked_ulimit < s_const.gibibyte // 2:
+            logger.warning(
+                'Operating system limit of maximum amount of locked memory (currently %d) is \n'
+                'too low for optimal performance.', locked_ulimit)
+
+        logger.debug('memory locking thread started')
+
+        # Note:  available might be larger than max_total in a container
+        max_total = s_thisplat.getTotalMemory()
+        available = s_thisplat.getAvailableMemory()
+
+        PAGESIZE = 4096
+        max_to_lock = (min(locked_ulimit,
+                           int(max_total * MAX_TOTAL_PERCENT),
+                           int(available * MAX_TOTAL_PERCENT)) // PAGESIZE) * PAGESIZE
+
+        self.max_could_lock = max_to_lock
+
+        path = self.path.absolute() / 'data.mdb'  # Path to the file that gets mapped
+        fh = open(path, 'r+b')
+        fileno = fh.fileno()
+
+        prev_memstart = 0  # The last start of file mapping, used to keep track when the mapping moves
+        prev_memend = 0  # The last end of the file mapping, so we can start from there
+
+        # Avoid spamming messages
+        first_end = True
+        limit_warned = False
+        self.locking_memory = True
+
+        self.resizeevent.set()
```

Comment:
Is there a race here vs the async thread?  Perhaps the event should be set to it's default set state before this thread is fired?

Reply:
self.resizeevent is thread safe. The event's "default" state is unset. It is OK if there are extra signals, since the loop will do nothing.
The reason I set it here is so it is obvious the loop will go around at least once before pending.